### PR TITLE
[Transaction Service Status] Batch status and memo writes to DB.

### PR DESCRIPTION
**Pull request by fkouteib**
#### Problem
Transaction status service issues individual writes to DB backend for each transaction memo, and each account/pubkey update within a transaction. This seems inefficient at least from CPU execution time perspective, even if the DB backend is aggregating and batching IO updates to disk.

#### Summary of Changes
* Batch transaction status writes before writing them to DB backend.
* Batch transaction memo writes before writing them to DB backend.
* Perform batching at transaction batch level (typically 64 tx).
* Add benches and unit tests.

### CODE REVIEW

**Comment by fkouteib** [comment]
This data was collected using a synthetic workload, from an internal test, and a 2-node local cluster setup running on a single physical node (Linux). This test runs in about a minute. The transaction mix is largely test transactions that have 35 account pubkeys including the fee payer [these land in TSS in 64 tx batches] and single vote transactions [they land in a batch of 1]. This was measured execution time for the match segment that handles transaction status batches in transaction status service.
 
| Stat | Baseline (exec time in us) |  batched (exec time in us) |
| ------------- | ------------- |------------- |
| min | 22 | 16 |
| max | 226,770 | 49,679  |
| mean | 517.01  | 186.23  |
| median | 233  | 87  |
| std dev | 2581.53  | 795.16  |
| datapoints | 112,066 | 124,027  |

I also ran the same synthetic workload and internal test on a small multi-node distributed private test cluster, and compared high level disk io metrics for num sectors written and num writes completed and the ratio were pretty close, and seemed to me the writes must generally be 128k commands based on the ratios averaged out.

**Comment by bw-solana** [comment]
> disk io metrics for num sectors written and num writes completed and the ratio were pretty close

This is interesting given the high level timings seem significantly better after this change. Do we think this is just due to reduction in CPU time from merging writes internally?

File: ledger/src/blockstore.rs
Comment (bw-solana) [inline ledger/src/blockstore.rs]: I'm thinking it might be better to split this into two APIs. One for the existing behavior and one for the new behavior (called something like `transaction_status_add_to_batch`)

File: ledger/src/blockstore.rs
Comment (bw-solana) [inline ledger/src/blockstore.rs]: Same thing for this API

File: ledger/src/blockstore.rs
Comment (bw-solana) [inline ledger/src/blockstore.rs]: We should probably have a unit test that confirms the new batching behavior

**Review by bw-solana — COMMENTED** [review]
Left a couple of suggestions.

I'm also wondering if we have an existing bencher for some of these operations. If not, might be nice to add one and confirm it shows benefits relative to the unbatched behavior.

**Comment by bw-solana** [comment]
@lijunwangs - I'm thinking you're the best person to review these changes while Tyera is out. Let me know if there is a better candidate

**Comment by lijunwangs** [comment]
> This data was collected using a synthetic workload, from an internal test, and a 2-node local cluster setup running on a single physical node (Linux). This test runs in about a minute. The transaction mix is largely test transactions that have 35 account pubkeys including the fee payer [these land in TSS in 64 tx batches] and single vote transactions [they land in a batch of 1]. This was measured execution time for the match segment that handles transaction status batches in transaction status service.
> 
> Stat	Baseline (exec time in us)	batched (exec time in us)
> min	22	16
> max	226,770	49,679
> mean	517.01	186.23
> median	233	87
> std dev	2581.53	795.16
> datapoints	112,066	124,027
> I also ran the same synthetic workload and internal test on a small multi-node distributed private test cluster, and compared high level disk io metrics for num sectors written and num writes completed and the ratio were pretty close, and seemed to me the writes must generally be 128k commands based on the ratios averaged out.

What is the exact metrics being used for this data?

File: ledger/src/blockstore.rs
Comment (lijunwangs) [reply ledger/src/blockstore.rs]: Agreed. Let's create an overload. Mixing single and batch is complicated.

**Comment by fkouteib** [comment]
> What is the exact metrics being used for this data?

@lijunwangs It's not an official metric that uploads to the metrics db, but I posted the debug code used for timing it [here](https://github.com/anza-xyz/agave/pull/3059).

File: ledger/src/blockstore.rs
Comment (fkouteib) [reply ledger/src/blockstore.rs]: Added.

**Comment by fkouteib** [comment]
> I'm also wondering if we have an existing bencher for some of these operations. If not, might be nice to add one and confirm it shows benefits relative to the unbatched behavior.

@bw-solana Added new benches.  To enable useful comparison, I used processing 64 tx per iteration, aligning to typical tx batch size.  Results on my Linux dev box:
```
test bench_write_transaction_memos         ... bench:     208,695.29 ns/iter (+/- 3,591.58)
test bench_add_transaction_memos_to_batch  ... bench:      78,976.33 ns/iter (+/- 14,641.83)

test bench_write_transaction_status        ... bench:   1,059,234.45 ns/iter (+/- 28,127.55)
test bench_add_transaction_status_to_batch ... bench:     507,835.53 ns/iter (+/- 28,981.67)
```

**Comment by bw-solana** [comment]
> > I'm also wondering if we have an existing bencher for some of these operations. If not, might be nice to add one and confirm it shows benefits relative to the unbatched behavior.
> 
> @bw-solana Added new benches. To enable useful comparison, I used processing 64 tx per iteration, aligning to typical tx batch size. Results on my Linux dev box:
> 
> ```
> test bench_write_transaction_memos         ... bench:     208,695.29 ns/iter (+/- 3,591.58)
> test bench_add_transaction_memos_to_batch  ... bench:      78,976.33 ns/iter (+/- 14,641.83)
> 
> test bench_write_transaction_status        ... bench:   1,059,234.45 ns/iter (+/- 28,127.55)
> test bench_add_transaction_status_to_batch ... bench:     507,835.53 ns/iter (+/- 28,981.67)
> ```

If I'm parsing this properly, looks like a 2-3x speedup from batching. Nice!

File: ledger/benches/blockstore.rs
Comment (bw-solana) [inline ledger/benches/blockstore.rs]: nit: maybe s/i/slot to make it more clear what this is

File: ledger/src/blockstore.rs
Comment (bw-solana) [inline ledger/src/blockstore.rs]: I'm leaning towards getting rid of this function (or making it private) and replacing with:

1. get_write_batch
2. write_batch

What do you think?

**Review by bw-solana — COMMENTED** [review]
Looks pretty good. Couple minor questions/suggestions

File: ledger/src/blockstore.rs
Comment (fkouteib) [reply ledger/src/blockstore.rs]: Cleaner. I like it. Fixed.

**Review by bw-solana — DISMISSED** [review]
LGTM. But would be good to get @lijunwangs approval before merging

**Comment by lijunwangs** [comment]
> @lijunwangs - I'm thinking you're the best person to review these changes while Tyera is out. Let me know if there is a better candidate

Taking another look.

File: ledger/src/blockstore.rs
Comment (lijunwangs) [inline ledger/src/blockstore.rs:2944]:
```suggestion
    #[inline]
    fn write_transaction_status_helper<'a, F>(
        &self,
        slot: Slot,
        signature: Signature,
        keys_with_writable: impl Iterator<Item = (&'a Pubkey, bool)>,
        status: TransactionStatusMeta,
        transaction_index: usize,
        mut write_fn: F,
    ) -> Result<()>
    where
        F: FnMut(&Pubkey, Slot, u32, Signature, bool) -> Result<()>,
    {
        let status = status.into();
        let transaction_index = u32::try_from(transaction_index)
            .map_err(|_| BlockstoreError::TransactionIndexOverflow)?;
        self.transaction_status_cf
            .put_protobuf((signature, slot), &status)?;

        for (address, writeable) in keys_with_writable {
            write_fn(address, slot, transaction_index, signature, writeable)?;
        }

        Ok(())
    }

    pub fn write_transaction_status<'a>(
        &self,
        slot: Slot,
        signature: Signature,
        keys_with_writable: impl Iterator<Item = (&'a Pubkey, bool)>,
        status: TransactionStatusMeta,
        transaction_index: usize,
    ) -> Result<()> {
        self.write_transaction_status_helper(
            slot,
            signature,
            keys_with_writable,
            status,
            transaction_index,
            |address, slot, tx_index, signature, writeable| {
                self.address_signatures_cf.put(
                    (*address, slot, tx_index, signature),
                    &AddressSignatureMeta { writeable },
                )
            },
        )
    }

    pub fn add_transaction_status_to_batch<'a>(
        &self,
        slot: Slot,
        signature: Signature,
        keys_with_writable: impl Iterator<Item = (&'a Pubkey, bool)>,
        status: TransactionStatusMeta,
        transaction_index: usize,
        db_write_batch: &mut WriteBatch<'_>,
    ) -> Result<()> {
        self.write_transaction_status_helper(
            slot,
            signature,
            keys_with_writable,
            status,
            transaction_index,
            |address, slot, tx_index, signature, writeable| {
                db_write_batch.put::<cf::AddressSignatures>(
                    (*address, slot, tx_index, signature),
                    &AddressSignatureMeta { writeable },
                )
            },
        )
    }
```

Comment (lijunwangs) [inline ledger/src/blockstore.rs:2944]: These two functions are almost identical except for the last write part -- even that has identical signature. Let's use a helper to share the code.

**Review by lijunwangs — CHANGES_REQUESTED** [review]

File: ledger/src/blockstore.rs
Comment (fkouteib) [reply ledger/src/blockstore.rs:2944]: Thx for the suggested fix. Accepted and resolved.

**Review by lijunwangs — APPROVED** [review]
looks good. Thanks for the improvements!

**Comment by steviez** [comment]
Nice, definitely a good change and inline with other blockstore/blockstore-adjacent code (such as shred insertion) that tries to batch up as much as possible.

Another possible optimization could be draining the channel in each invocation to `write_transaction_status_batch()` and batching those together. Hypothetically, TransactionStatusService (TSS) should be consuming at a similar rate that replay is producing. However, there are a number of things that could happen that could make TSS momentarily lag and allow several `TransactionStatusMessage`'s to accrue in the channel. We could probably just add a dumb log that that spits out `receiver.len()` and see if/how often that value is `> 1`.

**Pull request merged**
